packages/auth/src/interaction/service.ts

```diff
@@ -0,0 +1,142 @@
+import { v4 } from 'uuid'
+import { Transaction, TransactionOrKnex } from 'objection'
+
+import { IAppConfig } from '../config/app'
+import { BaseService } from '../shared/baseService'
+import { generateNonce } from '../shared/utils'
+import {
+  Interaction,
+  InteractionState,
+  InteractionWithGrant,
+  isInteractionWithGrant
+} from './model'
+import { GrantService } from '../grant/service'
+
+export interface InteractionService {
+  getInteractionByGrant(grantId: string): Promise<InteractionWithGrant | void>
+  getBySession(id: string, nonce: string): Promise<InteractionWithGrant | void>
+  getByRef(ref: string): Promise<InteractionWithGrant | void>
+  create(grantId: string, trx?: Transaction): Promise<Interaction>
+  approve(id: string): Promise<Interaction | void>
+  deny(id: string): Promise<Interaction | void>
+}
+
+interface ServiceDependencies extends BaseService {
+  grantService: GrantService
+  knex: TransactionOrKnex
+  config: IAppConfig
+}
+
+export async function createInteractionService({
+  grantService,
+  logger,
+  knex,
+  config
+}: ServiceDependencies): Promise<InteractionService> {
+  const log = logger.child({
+    service: 'InteractionService'
+  })
+  const deps: ServiceDependencies = {
+    grantService,
+    logger: log,
+    knex,
+    config
+  }
+  return {
+    getInteractionByGrant: (grantId: string) => getInteractionByGrant(grantId),
+    getBySession: (id: string, nonce: string) => getBySession(id, nonce),
+    getByRef: (ref: string) => getByRef(ref),
+    create: (grantId: string, trx?: Transaction) => create(deps, grantId, trx),
+    approve: (id: string) => approve(id),
+    deny: (id: string) => deny(id)
+  }
+}
+
+async function getInteractionByGrant(
+  grantId: string
+): Promise<InteractionWithGrant | void> {
+  const interaction = await Interaction.query()
+    .findOne({
+      grantId
+    })
+    .withGraphFetched('grant')
+
+  if (!interaction || !isInteractionWithGrant(interaction)) {
+    return undefined
+  }
+
+  return interaction
+}
+
+async function create(
+  deps: ServiceDependencies,
+  grantId: string,
+  trx?: Transaction
+): Promise<Interaction> {
+  const { knex, config } = deps
+  const interactionTrx = trx || (await Interaction.startTransaction(knex))
+
+  try {
+    const interaction = await Interaction.query(interactionTrx).insert({
```

Comment:
```suggestion
  try {
    const interaction = await Interaction.query(trx || deps.knex).insert({
```
would it be sufficient without `startTransaction`, given that it's just a single insert here?

Reply:
The way I've implemented the flow of grant creation, interactions are created [in tandem with grants](https://github.com/interledger/rafiki/blob/b2eea04995e1b2b9bf3b9cebb382aa83ff5bc7b2/packages/auth/src/grant/routes.ts#L165-L176) if they require one, so I added a transaction in case the corresponding grant creation fails. 

There might be a case for creating the interaction during the `start interaction` step rather than the `create grant` step.